### PR TITLE
Schedule with GitHub

### PR DESCRIPTION
**Pull request by naterush**
# Description

This PR completes the scheduling spec.

The main thing missing is that it doesn't disable this action in the case of custom edits or custom imports. We can / should do this.

# Testing

Use the feature! 

# Documentation

See documentation on using this feature here: https://docs.trymito.io/~/changes/y5Dq8XWydZBqGThnowWW/how-to/scheduling-your-automation

(Feel free to improve wording in the docs as well).

### CODE REVIEW

File: mitosheet/mitosheet/api/api.py [hunk 1]
```diff
@@ -49,7 +52,7 @@
 # so the API calls are handled in the main thread, to make printing easy.
 # In newer versions of JupyterLab, to see these print statements:
 # View > Show Log Console > in the console set Log Level to Debug
-THREADED = True
+THREADED = False
```
Comment (naterush) [inline mitosheet/mitosheet/api/api.py:55]: Remove

File: mitosheet/mitosheet/api/get_pr_url_of_new_pr.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+from mitosheet.types import StepsManagerType
+
+
+def get_pr_url_of_new_pr(params: Dict[str, Any], steps_manager: StepsManagerType) -> Union[str, Dict[str, str]]:
+    automation_name = params['automation_name']
+    automation_description = params['automation_description']
+    schedule = params['schedule']
+
+    # TODO: make this better
```
Comment (naterush) [inline mitosheet/mitosheet/api/get_pr_url_of_new_pr.py:14]: Remove this

File: mitosheet/mitosheet/api/get_pr_url_of_new_pr.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        )
+    except Exception as e:
+        return {
+            'error': str(e)
+        }
+
+    return pr_url
+
+
+"""
```
Comment (naterush) [inline mitosheet/mitosheet/api/get_pr_url_of_new_pr.py:43]: Remove

File: mitosheet/mitosheet/scheduling/schedule_utils.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
+import datetime
+import os
+import sys
+from typing import Any, Dict, Tuple
+from mitosheet.transpiler.transpile import transpile
+
+from mitosheet.types import StepsManagerType
+
+def get_path_of_import_param_for_automation(
```
Comment (naterush) [inline mitosheet/mitosheet/scheduling/schedule_utils.py:9]: Doc string

Comment (naterush) [reply mitosheet/mitosheet/scheduling/schedule_utils.py:9]: For all functions

File: mitosheet/mitosheet/scheduling/schedule_utils.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+        raise Exception(f"Invalid schedule type: {schedule['type']}")
+    
+def get_automation_code(
+        steps_manager: StepsManagerType,
+        automation_name: str,
+        automation_base_folder: str,
+    ) -> str:
+
+    # First, we get the function name
+    # TODO: improve this
```
Comment (naterush) [inline mitosheet/mitosheet/scheduling/schedule_utils.py:82]: Update all these things... we should have:

1. `Automation Name`
2. `Automation Name as Python Variable`
3. `Automation Name as Folder` -- maybe this is just (2)?

File: mitosheet/mitosheet/tests/scheduling/test_schedule_utils.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+from mitosheet.scheduling.schedule_utils import get_automation_code
+
+def test_get_automation_code_with_no_params():
+    mito = create_mito_wrapper()
+    automation_code = get_automation_code(
+        mito.mito_backend.steps_manager,
+        'automation_name',
+        'automation_name',
+    )
+    print(automation_code)
```
Comment (naterush) [inline mitosheet/mitosheet/tests/scheduling/test_schedule_utils.py:12]: Remove

File: mitosheet/mitosheet/tests/scheduling/test_schedule_utils.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+# Create a folder for this run
+export_time = sys.argv[1]
+os.makedirs(f"automation_name/runs/{export_time}")
+
+# Run the automation
+file_name_import_csv_0 = r"automation_name/data/input.csv"
+file_name_export_csv_0 = f"automation_name/runs/{export_time}/file_name_export_csv_0.csv"
+file_name_export_excel_0 = f"automation_name/runs/{export_time}/file_name_export_excel_0.xlsx"
+automation_name(file_name_import_csv_0, file_name_export_csv_0, file_name_export_excel_0)"""
```
Comment (naterush) [inline mitosheet/mitosheet/tests/scheduling/test_schedule_utils.py:98]: Tests:
1. Test the cron string thing
2. Tests that we get all the files we expect (and that they are non-empty)

File: mitosheet/mitosheet/transpiler/transpile_utils.py [hunk 8]
```diff
@@ -335,11 +340,23 @@ def get_imports_for_custom_python_code(code: List[str], steps_manager: StepsMana
...
+            #else:
+            #    inlined_functions.append(inspect.getsource(func))
+                
+
     import_strings = [
         f"from {module_name} import {', '.join(function_names)}"
         for module_name, function_names in import_map.items()
     ]
 
+    import_strings.extend(inlined_functions)
```
Comment (naterush) [inline mitosheet/mitosheet/transpiler/transpile_utils.py:358]: Remove

File: mitosheet/src/mito/components/elements/AutomationSchedulePicker.tsx [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+                            value={schedule.dayOfMonth + ''}
+                            onChange={(newValue) => {
+                                props.setSchedule({
+                                    ...schedule,
+                                    dayOfMonth: parseInt(newValue),
+                                })
+                            }}
+                        >
+                            {Array.from(Array(31).keys()).map((day) => {
+                                return <DropdownItem title={day + 1 + ''} id={day + 1 + ''}/>
```
Comment (naterush) [inline mitosheet/src/mito/components/elements/AutomationSchedulePicker.tsx:124]: Key

File: mitosheet/src/mito/components/icons/ScheduleIcon.tsx [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+import React from 'react';
+
+
+const ScheduleIcon = (): JSX.Element => {
+
+    return (
+        <svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
+            <circle cx="8" cy="8" r="7.5" fill="var(--mito-background-off)" stroke="var(--mito-text)"/>
+            <line x1="8" y1="8.875" x2="12" y2="8.875" stroke="var(--mito-highlight)" stroke-width="1.25"/>
```
Comment (naterush) [inline mitosheet/src/mito/components/icons/ScheduleIcon.tsx:11]: lints

File: mitosheet/src/mito/components/taskpanes/GithubSchedule/GithubScheduleTaskpane.tsx [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+import TextButton from "../../elements/TextButton";
+import DefaultTaskpane from "../DefaultTaskpane/DefaultTaskpane";
+import DefaultTaskpaneBody from "../DefaultTaskpane/DefaultTaskpaneBody";
+import DefaultTaskpaneFooter from "../DefaultTaskpane/DefaultTaskpaneFooter";
+import DefaultTaskpaneHeader from "../DefaultTaskpane/DefaultTaskpaneHeader";
+import Input from "../../elements/Input";
+import Row from "../../layout/Row";
+import Col from "../../layout/Col";
+import LabelAndTooltip from "../../elements/LabelAndTooltip";
```
Comment (naterush) [inline mitosheet/src/mito/components/taskpanes/GithubSchedule/GithubScheduleTaskpane.tsx:13]: Remove unused import

**Review by naterush — COMMENTED** [review]
Cleanup!

**Comment by aarondr77** [comment]
My automations are creating new GitHub branches, but I'm not getting any confirmation 

https://github.com/mito-ds/mito/assets/18709905/112ebda5-cd85-4c83-be3f-86829090a35a

**Comment by aarondr77** [comment]
If I use a custom import, it's letting me schedule.

File: mitosheet/mitosheet/scheduling/github.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+        pr_description: str, 
+        new_files: Dict[str, str]
+    ) -> str:
+    """
+    Create a new GitHub PR with the given parameters.
+    """
+
+    token = os.environ.get('GITHUB_TOKEN')
+    if not token:
+        raise Exception("GitHub token not found in environment variables. See docs here: TODO: insert link")
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/scheduling/github.py:22]: These will be our docs, once they are published.

File: mitosheet/mitosheet/scheduling/schedule_utils.py [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+        export_folder_path = f'{automation_base_folder}/runs/{{export_time}}'
+        export_folder_setup.append(f'os.makedirs(f"{export_folder_path}")')
+
+    # Then, we actually call the function
+    function_call = f'{function_name}('
+    for param_metadata in steps_manager.param_metadata:
+        function_call += f'{param_metadata["name"]}, '
+    if len(steps_manager.param_metadata) > 0:
+        function_call = function_call[:-2]
+    function_call += ')'
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/scheduling/schedule_utils.py:175]: Let's do this refactor when we have need of this code. It's well encapsulated for now.

File: mitosheet/mitosheet/scheduling/schedule_utils.py [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+def get_requirements_txt_file() -> str:
+    """
+    Gets the requirements.txt file that should be used for the automation -- freezing the current 
+    dependency versions of the big dependencies.
+
+    TODO: in the future, we can handle other dependencies used in this automation (how to detect 
+    automatically?) -- are there every any others?
+    """
+    from mitosheet import __version__ as mitosheet_version
+    if mitosheet_version == '0.3.131': # Handle development version
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/scheduling/schedule_utils.py:202]: This is what's hard-coded in our development enviornemtns, per the comment. Will make it more clear

File: mitosheet/mitosheet/transpiler/transpile_utils.py [hunk 15]
```diff
@@ -335,6 +339,16 @@ def get_imports_for_custom_python_code(code: List[str], steps_manager: StepsMana
 
                 import_map[module_name] = imports_from_module
 
+            # Otherwise, we are in JupyerLab, and if we're being asked to import
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/transpiler/transpile_utils.py:342]: Yeah! This should totally be in our frontend tests.

**Comment by naterush** [comment]
Merging when lints pass

**Pull request merged**